<a href="https://colab.research.google.com/github/amonguscline/ML-Data-analysis/blob/main/projects/Machine_Learning_via_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
from sklearn.linear_model import LogisticRegression

In [ ]:
df=pd.read_csv("/content/sample_data/california_housing_test.csv")

bayarea=pd.DataFrame()
bayarea["longitude"]=df.iloc[:,0].map(lambda s:1 if s>-122.644 and s<-121.587 else 0)
bayarea["latitude"]=df.iloc[:,1].map(lambda s:1 if s>37.189 and s<38.203 else 0)
def combineRows(a,b):
    return a and b
df["in_bay_area"]=bayarea.iloc[:,0:2].apply(lambda row:combineRows(row["longitude"],row["latitude"]),axis=1)
X=df.iloc[:,2:-1]
y=df["in_bay_area"]
model = LogisticRegression(max_iter=1000)

model.fit(X,y)
model.score(X,y)
sample=df.sample(10)
model.predict_proba(sample.iloc[:,2:-1])

array([[0.91456113, 0.08543887],
       [0.64156354, 0.35843646],
       [0.88965789, 0.11034211],
       [0.54798454, 0.45201546],
       [0.53054255, 0.46945745],
       [0.84350467, 0.15649533],
       [0.46310749, 0.53689251],
       [0.81429199, 0.18570801],
       [0.74484166, 0.25515834],
       [0.79335268, 0.20664732]])

In [ ]:
sample.iloc[:,2:]

,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,in_bay_area
2138,34.0,2200.0,493.0,1243.0,431.0,1.8514,66500.0,0
2455,45.0,1685.0,321.0,815.0,314.0,4.2955,309700.0,1
155,16.0,3907.0,671.0,1759.0,663.0,3.1776,172600.0,0
555,20.0,10722.0,1617.0,3731.0,1511.0,9.7449,500001.0,0
1285,52.0,2074.0,401.0,1136.0,409.0,4.7703,331000.0,1
2590,37.0,1801.0,422.0,1564.0,425.0,3.1597,133000.0,0
2850,35.0,2063.0,266.0,676.0,252.0,8.5294,500001.0,1
1612,21.0,3833.0,595.0,1645.0,589.0,4.6250,273500.0,0
974,43.0,952.0,209.0,392.0,210.0,2.1635,244200.0,0
959,35.0,2985.0,474.0,1614.0,453.0,5.4631,225600.0,0


In [ ]:
df["in_bay_area"].describe()

,in_bay_area
count,3000.000000
mean,0.189000
std,0.391574
min,0.000000
25%,0.000000
50%,0.000000
75%,0.000000
max,1.000000


In [ ]:
model.score(X,y)

0.804

The model scored really well, but I think that this is a good example of how the score of the model doesn't exactly reflect how good or useful the model actually is. since the data isn't split very well (only about 19% of the houses in the data were in the bay area), the model can get pretty accurate by always guessing 0, and the math actually lines up really well with this (with an .804 scoring accuracy). I think the way that you would best improve this model could be by including los angeles (perhaps this will make the model think more about what it choses because the split is more even), but that's a lot of work